Skip query cache for in_batches and friends

eugeneius opened the pull request:
Fixes https://github.com/rails/rails/issues/28646.

The `find_each`, `find_in_batches` and `in_batches` APIs usually operate on large numbers of records, where it's preferable not to load them all into memory at once.

If the query cache is enabled, it will hold onto the query results until the end of the execution context (request/job), which means the memory used is still proportional to the total number of records. These queries are typically not repeated, so the query cache isn't desirable here.

r? @matthewd

eugeneius committed:
Skip query cache for in_batches and friends

The `find_each`, `find_in_batches` and `in_batches` APIs usually operate
on large numbers of records, where it's preferable not to load them all
into memory at once.

If the query cache is enabled, it will hold onto the query results until
the end of the execution context (request/job), which means the memory
used is still proportional to the total number of records. These queries
are typically not repeated, so the query cache isn't desirable here.

matthewd commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -211,39 +211,41 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
+          raise ArgumentError.new("Primary key not included in the custom select clause") unless primary_key_offset
 
-        yield yielded_relation
+          yield yielded_relation
```
I think this might surprise people, by spreading the uncaching too far.

It rather feels like we might need to add a "don't cache this" flag to the relation itself.

eugeneius commented:
I've reworked this to avoid caching the batched records without disabling the query cache for the duration of the block, by adding a flag to the yielded relation as @matthewd suggested.

matthewd commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -209,7 +209,7 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
       relation = relation.reorder(batch_order).limit(batch_limit)
       relation = apply_limits(relation, start, finish)
-      batch_relation = relation
+      batch_relation = relation.uncached!
```
Mutation at the same time we copy it to a second variable seems confusing. Maybe move this call up to 210?

Actually, let's move it to its own line, so we can add a comment: `# Retaining the results in the query cache would undermine the point of batching`

I think that's necessary partly because the tests check it's uncached, but don't really illustrate the *reason* it is. And a comment seems easier than finding a way to make an assertion about query cache bloat. :smile:

matthewd commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -218,7 +218,7 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
           yielded_relation.load_records(records)
         else
-          ids = batch_relation.pluck(primary_key)
+          ids = uncached { batch_relation.pluck(primary_key) }
```
Let's chase `uncached_value` through `relation/{calcuations,finder_methods}.rb`.

We won't make it public API just now, but this still seems the best time to give it full coverage of queries the relation can make.

matthewd commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -657,7 +657,9 @@ def has_join_values?
+        end
 
         preload = preload_values
         preload += includes_values unless eager_loading?
```
I think the `preload` call just down ⬇️ there wants to be wrapped too. Don't want to fill the cache with `SELECT * FROM children WHERE parent_id IN (long, list, of, ids)` for each batch.

matthewd reviewed (commented):
This looks fantastic! ❤️ 💙 💜

eugeneius commented:
Thanks for the detailed feedback! I think I've addressed everything.

I renamed the method to `skip_query_cache!`, because I was nervous about potential confusion between a new undocumented-but-public `uncached!` method and [the existing `uncached` method](https://github.com/rails/rails/blob/v5.1.1/activerecord/lib/active_record/query_cache.rb#L17).

I also discovered that `#exists?` queries never hit the query cache on the MySQL and PostgreSQL adapters, because they both reimplement `select_rows` without going through `select_all`:

https://github.com/rails/rails/blob/v5.1.1/activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb#L18-L23
https://github.com/rails/rails/blob/v5.1.1/activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb#L28-L32

I'll look into fixing that problem, but I don't think it should necessarily block this change; I've skipped the `#exists?`-related tests on those adapters for now.

eugeneius commented:
https://github.com/rails/rails/pull/29454 fixes the query cache for `#exists?`; if it gets merged I'll rebase this and remove the `skip`s.

rafaelfranca commented:
Please update it

eugeneius commented:
Updated!

eugeneius commented:
@matthewd I think this is ready to go - would you mind taking another look?

nateberkopec commented:
Makes a lot of sense to me 👍

matthewd commented:
👌🏻
